uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php

```diff
@@ -135,4 +145,238 @@ public function testDeleteDeletesTheLeaveRequestItsBalanceChangesAndDates() {
 
     $this->fail("Expected to not find the LeaveRequest with {$leaveRequest->id}, but it was found");
   }
+
+  /**
+   * @expectedException RuntimeException
+   * @expectedExceptionMessage You are not allowed to create or update a leave request for this employee
+   */
+  public function testCreateThrowsAnExceptionWhenCurrentUserDoesNotHaveCreateAndUpdateLeaveRequestPermission() {
+    //logged in user has no permissions, also a contactID different from that of the logged in user is passed
+    $contactID = 2;
+    $this->getleaveRequestService()->create([
+      'type_id' => 1,
+      'contact_id' => $contactID,
+      'status_id' => 1,
+      'from_date' => CRM_Utils_Date::processDate('2016-01-04'),
+      'from_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'to_date' => CRM_Utils_Date::processDate('2016-01-10'),
+      'to_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+    ], false);
+  }
+
+  /**
+   * @expectedException RuntimeException
+   * @expectedExceptionMessage You can't create a Leave Request with this status
+   */
+  public function testCreateThrowsAnExceptionWhenTransitionStatusIsNotValidForNewLeaveRequest() {
+    //A leave contact trying to set a new leave request to approved status
+    $this->getLeaveRequestServiceWhenStatusTransitionIsNotAllowed()->create([
+      'type_id' => 1,
+      'contact_id' => $this->leaveContact,
+      'status_id' => $this->leaveRequestStatuses['Approved']['id'],
+      'from_date' => CRM_Utils_Date::processDate('2016-01-04'),
+      'from_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'to_date' => CRM_Utils_Date::processDate('2016-01-10'),
+      'to_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+    ], false);
+  }
+
+  public function testCreateThrowsAnExceptionWhenTransitionStatusIsNotValidWhenUpdatingLeaveRequestStatus() {
+    $leaveRequest = LeaveRequestFabricator::fabricateWithoutValidation([
+      'contact_id' => $this->leaveContact,
+      'type_id' => 1,
+      'from_date' => CRM_Utils_Date::processDate('2016-01-04'),
+      'from_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'to_date' => CRM_Utils_Date::processDate('2016-01-10'),
+      'to_date_type' => $this->getLeaveRequestDayTypes()['All Day']['value'],
+      'status_id' => $this->leaveRequestStatuses['More Information Requested']['id'],
+      'sequential' => 1,
+    ]);
```

Comment:
Its actually necessary here because the the `statusChanged` method of the leave request service is invoked behind the scenes and it needs a leave request to have existed before.